GUIDELINES.md

```diff
@@ -0,0 +1,135 @@
+
+## One-by-one steps
+
+* Create a module testing configuration in the test sources for each module in your project (e.g. for a module `app` create a file `AppModuleTestingConfiguration` with the following code: `val appModuleTestingConfiguration = moduleTestingConfiguration { ... }`; see below for more info on how to add dependencies to the configuration)
+  * Always use the same package structure for test sources, e.g. if you app resides in a module `app` and has a base package `org.myorg.product.app` then put the module test configuration in exactly in the same package
+  * If there is a class `org.myorg.product.app.view.LoginViewModel` then the test class `LoginViewModelTest` and `LoginViewModelSteps` would reside in the exact same package, too
+  * If you have a dedicated module for test sources you should still stick to the same package structure, e.g. if the module is `:app:test` _don't_ add `test` to the package structure for test sources
+  * If modules depend on each other the module testing configuration should do the same, e.g. `val appModuleTestingConfiguration = moduleTestingConfiguration(dependentConfig1, dependentConfig2, ...) { ... }`, this automatically imports all dependent configurations
+* Create a test class for the _feature under test_ (e.g. `LoginTest`), so the test is agnostic of the underlying structures (classes, models, views, data structures, ...)
+  * Should it be that a test really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelTest`)
+* Create a steps class for the feature under test (e.g. `LoginSteps`) which will contain all the `given`, `when` and `then` functions as well as dependency configuration and setup code
+  * The steps class should know about the setup of the test, so for example if the steps class tailored to test the integration of multiple classes, it should also be called something like `LoginIntegrationSteps` which tests a broad stack of classes (e.g. a `LoginViewModel` talking to an `AuthManager` and a `SessionStore`)
+  * Should it be that a steps class really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelSteps`)
+  * In case the steps class' aim is purely to offer and act on a fake implementation of a class, name it with a `FakeSteps` suffix (e.g. `BackendGatewayFakeSteps` or `SessionStoreFakeSteps`)
+* Start with the test
+  * In the test class you created add reference to the steps class with the `steps` function (e.g. `val sut by steps<LoginSteps>`) and call the value `sut` (system under test)
+  * Fill up the test with `@Test`-annotated test functions
+    * The test function can be named after a _scenario_ (e.g. `New user trying to log in`, `Weather condition is cloudy`)
+    * Only if necessary to distinguish test cases, also add more context like what's roughly expected (e.g. `New user trying to login - welcome to the app message shown`, `Existing user trying to login - welcome back message shown`)
```

Comment:
Not sure, if I agree there. Having long test names is not a bad thing. Helps to see what failed immediately what went wrong in a report. And this will also be looked over in Audits. The more descriptive the better.

Reply:
i'm not sure what auditors exactly look at, but i think overall comprehensiveness is more important. in the end the test function name won't give anyone any kind of security or satisfying insight into the test. also the argument that you exactly know what failes when you look at the test name is not 100% correct. it can help narrowing down the issue but won't replace going deeply into the code in order to understand the problem. on top of that, the exact circumstances can way more precisely be extracted from the stack trace of the failed test. the abstracted style of tests like we do it in sweetest helps most in making the test more comprehensive.

so given the fact you usually spend most time reading and modifying tests my focus is more on the distinctiveness of the test function name over completeness. i think we already do it like that in our code, so it's probably not about our different styles but about the proper wording in these guidelines. we could use e.g. instead of "you should do this or that" something of the form like "you don't have to..." and moreover "make sure it's descriptive enough to understand what's tested as fast as possible, meanwhile making it distinctive in order to be able to havigate through the test fast" and add some examples...

Reply:
> the argument that you exactly know what failes when you look at the test name is not 100% correct. 

That depends how the tests are written, I like this [article](https://proandroiddev.com/seven-principles-of-great-unit-tests-adapted-for-android-342515f98ef2) a lot. The section 2. Independent describe it well I think.

As said, examples would help to clarify the naming conventions.

Reply:
the respective paragraph in this article is not 100% conclusive for me as it doesn't give me an exact reason, but i think this is probably too much into detail for this guidelines. i'm considering removing that point...